flake8_rst/sourceblock.py

```diff
@@ -0,0 +1,201 @@
+
+        if len(source_blocks) == 1:
+            return source_blocks[0]
+        main_block = source_blocks[0]
+        boot_lines = main_block.boot_lines
+        source_lines = []
+        for source_block in source_blocks:
+            if boot_lines != source_block.boot_lines:
+                raise ValueError('You cannot merge SourceBlocks with different bootstraps!')
+            source_lines.extend(source_block.source_lines)
+        source_lines.sort(key=lambda line: line[LINENO])
+        return cls(boot_lines, source_lines, main_block.directive, main_block.language)
+
+    def __init__(self, boot_lines, source_lines, directive='', language='', roles=None):
+        self._boot_lines = boot_lines
+        self._source_lines = source_lines
+        self.directive = directive
+        self.language = language
+        self.roles = roles or {}
+        self.ignore_lines_with = DEFAULT_IGNORED_LINES
+        self.console_syntax = DEFAULT_CONSOLE_SYSNTAX
+
+        self.roles.setdefault('group', 'None' if directive != 'ipython' else directive)
+        if 'bootstrap' in self.roles:
+            self.roles['group'] = 'None'
+            self._boot_lines = SourceBlock.convert_bootstrap(self.roles['bootstrap'], split='; ')
+
+    @property
+    def boot_lines(self):
+        return self._boot_lines
+
+    @property
+    def source_lines(self):
+        return self._source_lines
+
+    @property
+    def source_block(self):
+        """Return code lines **without** bootstrap"""
+        return "".join((line[SOURCE] for line in self._source_lines))
+
+    @property
+    def complete_block(self):
+        """Return code lines **with** bootstrap"""
+        return "".join([line[SOURCE] for line in self._boot_lines + self._source_lines])
+
+    def get_code_line(self, lineno):
+        all_lines = self._boot_lines + self._source_lines
+        line = all_lines[lineno - 1]
+        return {'lineno': line[LINENO], 'indent': len(line[RAW]) - len(line[SOURCE]),
+                'source': line[SOURCE], 'raw_source': line[RAW]}
+
+    def find_blocks(self, expression):
+
+        src = self.source_block
+        for match in expression.finditer(src):
+            origin_code = str(match.group('code'))
+            line_start = src[:match.start()].count('\n') + match.group('before').count('\n')
+            source_slice = slice(line_start, line_start + len(origin_code.splitlines(True)))
+            directive = _match_default(match, 'directive', '')
+            language = _match_default(match, 'language', '')
+            roles = _extract_roles(match)
+
+            source_block = SourceBlock(self._boot_lines, self._source_lines[source_slice], directive=directive,
+                                       language=language, roles=roles)
+            source_block._remove_indentation()
+            yield source_block
+
+    def _remove_indentation(self):
+        expression = re.compile('(?P<indent>^ *).', re.MULTILINE)
```

Comment:
Also compile at func call

Reply:
I've now searched for `re.compile` within the file. All are outside now.